doc/nrf/libraries/networking/nrf_cloud_rest.rst

```diff
@@ -10,22 +10,77 @@ nRF Cloud REST
+This library exposes a number of functions that allow access to nRF Cloud REST API endpoints.
 
-Pass the REST context structure, along with any other required parameters, to the desired function.
-When the function returns, the context contains the status of the API call.
+When called, these functions initiate an HTTP request to the nRF Cloud REST backend.
+Various endpoints (documented `here <nRF Cloud REST API_>`_) are provided for such tasks as updating a device's shadow, sending and receiving device messages, interacting with location services, and so on.
+
+All of these functions operate within the context of a :c:struct:`nrf_cloud_rest_context` structure, which you must initialize and provide with your desired parameters.
+When each function returns, the context contains the status of the API call.
+
+Authentication
+**************
+
+Each HTTP/REST request must include a valid JSON Web Token, which serves as a proof of identity.
+These tokens are generated, usually, based on the `device credentials. <nRF Cloud Security_>`_, and must be passed to the ``auth`` parameter of the :c:struct:`nrf_cloud_rest_context` structure.
+JWT Tokens can be generated by calling :c:func:`nrf_cloud_jwt_generate` or using :ref:`lib_modem_jwt`.
+Alternatively, if :kconfig:`NRF_CLOUD_REST_AUTOGEN_JWT` is enabled (along with its dependencies), the nRF Cloud REST library will generate a JWT automatically if one is not provided.
+
+Socket Management and Reuse
+***************************
+
+The socket used for each HTTP request is determined by the ``connect_socket`` parameter of the :c:struct:`nrf_cloud_rest_context` structure.
+If a valid socket file descriptor is passed, this socket is used. If -1 is passed, a socket will be created automatically.
+
+The selected socket can be reused depending on the passed-in value of the ``keep_alive`` property of the :c:struct:`nrf_cloud_rest_context` structure:
+
+ * If ``keep_alive`` is set to false, then the selected socket will be closed after each request, and the ``connect_socket`` property will be reset to -1.
+ * If ``keep_alive`` is set to true, the socket will remain open and ``connect_socket`` will remain unaltered, meaning the socket will be reused on the next API call.
+
+However, if ``keep_alive`` is set to true, care must be taken to ensure that the socket has not been closed externally (for example, due to inactivity) before sending each request.
+Otherwise, the request will be dropped.
+
+There are two ways a timed out socket can manifest:
+
+* The socket may be closed immediately on timeout, causing `-ENOTCONN` to be returned immediately by the next REST request function call, without reaching nRF Cloud backend.
+* The socket may be left open, allowing the next REST request to reach nRF Cloud Backend, but the request will be ignored and the socket will be closed gracefully (FIN,ACK) with null HTTP response.
```

Comment:
How do you know it reaches the backend?  It could have been closed by the carrier (without notice) due to a NAT timeout, in which case, the backend would never see it.  If the backend did receive it, it should process it and not return a null HTTP response.

Reply:
> How do you know it reaches the backend?
This is indicated by modem traces showing TLS alert response (with unknown content) originating from the endpoint IP. AFAIK only the backend could generate these encrypted alerts. Technically, the FIN,ACK also requires a valid sequence number which would be difficult for the carrier network to fake. The null HTTP response is a genuine response from back-end, and not generated at the carrier level.

> it should process it and not return a null HTTP response.

Officially, yes, all valid HTTP requests that reach an HTTP enpoint should receive valid HTTP responses. However AWS seems to violate this behavior if the socket being used has gone inactive, and instead gracefully closes the socket.

This does kind of depend on what is meant by backend. You could consider AWS as either the container for nRF Cloud Backend, or a necessary container that backend. It might be better to say "backend servers".

Reply:
We don't (or shouldn't) refer to the backend per se in NCS documentation.  I think all of this is handled by AWS -- I don't think we host or manage the HTTP server or its responses.  I wonder if this error condition is specifically because the JWT expired?

Reply:
> We don't (or shouldn't) refer to the backend per se in NCS documentation.
> We don't (or shouldn't) refer to the backend per se in NCS documentation. 
Ah, well, I definitely think we need to somehow explain that there are two totally different ways a timeout can manifest, and that it's essentially random which of the two occurs. I'm not sure there's a way to explain what is happening without referring to the backend somehow. Would referring to it indirectly be acceptable? 

> I wonder if this error condition is specifically because the JWT expired?
Definitely not caused by the JWTs -- they are all valid -- a new one is generated for each and every request, even if the socket is being reused.

Reply:
I don't think that saying "nRF Cloud backend" has any useful meaning or distinction to the customer from simply "nRF Cloud."  Such a distinction does not help them fix whatever is wrong.

Reply:
You're right, that is a vague term -- what about nRF Cloud REST endpoint server? Or just REST endpoint server? Or do you still think just saying "nRF Cloud" would be best?

Reply:
IMO to a customer they are one and the same.